### PR TITLE
Add a basic KV cache handler

### PR DESCRIPTION
**Pull request by vicb**

### CODE REVIEW

File: packages/cloudflare/package.json [hunk 1]
```diff
@@ -29,9 +29,11 @@
   },
   "homepage": "https://github.com/opennextjs/opennextjs-cloudflare",
   "devDependencies": {
+    "@cloudflare/workers-types": "catalog:",
     "@types/node": "catalog:",
     "esbuild": "catalog:",
     "glob": "catalog:",
+    "next": "catalog:",
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/package.json:36]: Yes, for the `CacheHandler`

File: packages/cloudflare/src/build/build-worker.ts [hunk 2]
```diff
@@ -1,47 +1,66 @@
...
 /** The directory containing the Cloudflare template files. */
-const templateSrcDir = path.join(path.dirname(fileURLToPath(import.meta.url)), "templates");
+const packageDir = path.dirname(fileURLToPath(import.meta.url));
 
 /**
  * Using the Next.js build output in the `.next` directory builds a workerd compatible output
  *
  * @param outputDir the directory where to save the output
- * @param nextjsAppPaths
+ * @param config
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/build/build-worker.ts:26]: Agreed

File: packages/cloudflare/src/build/build-worker.ts [hunk 3]
```diff
@@ -98,22 +117,21 @@ export async function buildWorker(
...
 import { Readable } from 'node:stream';
-globalThis.fetch.__nextPatched = isPatchedAlready;
 fetch = globalThis.fetch;
 const CustomRequest = class extends globalThis.Request {
   constructor(input, init) {
-    console.log("CustomRequest", input);
     if (init) {
       delete init.cache;
       if (init.body?.__node_stream__ === true) {
+        // https://github.com/cloudflare/workerd/issues/2746
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/build/build-worker.ts:134]: The hope is that this is getting resolved before we have time to add a comment ;)

File: packages/cloudflare/src/build/build.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import { rm } from "node:fs/promises";
 import { buildNextjsApp } from "./build-next-app";
 import { buildWorker } from "./build-worker";
-import { getNextjsAppPaths } from "../nextjs-paths";
+import { containsDotNextDir, getConfig } from "../config";
 import { cpSync } from "node:fs";
-import { resolve } from "node:path";
+import path from "node:path";
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/build/build.ts:6]: https://github.com/opennextjs/opennextjs-cloudflare/issues/30

File: packages/cloudflare/src/config.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
+import { readdirSync, statSync } from "node:fs";
+import path, { relative } from "node:path";
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/config.ts:2]: https://github.com/opennextjs/opennextjs-cloudflare/issues/30

File: packages/cloudflare/src/config.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+const PACKAGE_NAME = "@opennextjs/cloudflare";
+
+// Make this user configurable
+const UserConfig = {
+  cache: {
+    bindingName: "NEXT_CACHE_WORKERS_KV",
+  },
+};
+
+export type Config = {
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/config.ts:13]: +1 for a better name.
We can change to something more consistent with other open next adapters later.

File: packages/cloudflare/src/config.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    // Path to the `.next` directory specific to the standalone application
+    standaloneAppDotNext: string;
+    // Path to the `server` directory specific to the standalone application
+    standaloneAppServer: string;
+    // Package in the standalone node_modules
+    internalPackage: string;
+  };
+
+  cache: {
+    bindingName: string;
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/config.ts:32]: Done

File: packages/cloudflare/tsup.config.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { cp } from "fs/promises";
 import { defineConfig } from "tsup";
 
 export default defineConfig({
-  entry: ["src/index.ts"],
+  entry: ["src/index.ts", "src/cache-handler.ts"],
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/tsup.config.ts:5]: Let's updated when needed

File: packages/cloudflare/src/build/patches/investigated/copy-package.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import path from "node:path";
+import { Config } from "../../../config";
+import { cpSync } from "node:fs";
+
+/**
+ * Copy the builder package in the standalone node_modules folder.
+ */
+export function copyPackage(srcDir: string, config: Config) {
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/build/patches/investigated/copy-package.ts:8]: It's used to access the cache handler.

File: packages/cloudflare/src/config.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  };
+
+  // Internal name for the copy of the package
+  internalPackageName: string;
+};
+
+/**
+ * Computes the configuration.
+ *
+ * @param root folder containing built next app.
```
[the remark this answers is not in the document]
Comment (vicb) [reply packages/cloudflare/src/config.ts]: fixed

**Comment by vicb** [comment]
Thanks a lot for the review @dario-piotrowicz 
I have skipped some of the minor comments to make progress on the cache implementation.
We'll revisit later.

**Comment by dario-piotrowicz** [comment]
mine was only a partial review (sorry for not making it clear) I hadn't yet reviewed the most important file 🙈 (the cache impl one)

but yeah we can iterate over it later 🙂

**Pull request merged**